[Hub Generated] Review request for Microsoft.Security to add version preview/2023-09-01-preview

ivadim opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/ivadim-security-Microsoft.Security-devops-2023-09-01-preview...main/).
## ARM API Information (Control Plane)

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month. **Guestimate - July 2023**
  4. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month. **Guestimate - September 2023**
  5. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version). Note that this doesn't apply if you are trying to merge a PR that was previously in the private repository.

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/overview_breakingchanges).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](http://aka.ms/aprwiki))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

ivadim committed:
Adds base for updating Microsoft.Security from version preview/2023-05-01-preview to version 2023-09-01-preview

ivadim committed:
Updates readme

ivadim committed:
Updates API version in new specs and examples

ivadim committed:
Add initial version of SecurityConnecto DevOps swagger specification

JohnathonMohr committed:
Add examples, fix prettier issues

JohnathonMohr committed:
Update details in request/response example objects

JohnathonMohr committed:
Remove ids from AzDO projects/repos in request

JohnathonMohr committed:
Updating Onboarding State name

JohnathonMohr committed:
Adding back newline

JohnathonMohr commented:
/azp run unifiedPipeline

JohnathonMohr committed:
Fine tuning examples with onboardingState, adding available inventory examples

kazrael2119 commented:
/zap run

kazrael2119 commented:
/azp run

ms-henglu commented on specification/security/resource-manager/Microsoft.Security/preview/2023-09-01-preview/securityConnectorsDevOps.json:
```diff
@@ -0,0 +1,3368 @@
+        ],
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/SubscriptionIdParameter"
```
please always use the latest, current is "v5". please also update other places

ivadim replied:
Updated all references to v5

ms-henglu commented on specification/security/resource-manager/Microsoft.Security/preview/2023-09-01-preview/securityConnectorsDevOps.json:
```diff
@@ -0,0 +1,3368 @@
+          }
+        ],
+        "responses": {
+          "200": {
```
Since the delete operaiton is a long running operation, it should not have the 200 status code.
refs: https://github.com/Azure/azure-openapi-validator/blob/main/docs/delete-response-codes.md

ms-henglu commented on specification/security/resource-manager/Microsoft.Security/preview/2023-09-01-preview/securityConnectorsDevOps.json:
```diff
@@ -0,0 +1,3368 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Security/securityConnectors/{securityConnectorName}/devops": {
```
Consider adding the x-ms-pageable extension.

refs: https://github.com/Azure/azure-openapi-validator/blob/main/docs/pageable-operation.md

ivadim replied:
Added "x-ms-pageable"

ms-henglu commented on specification/security/resource-manager/Microsoft.Security/preview/2023-09-01-preview/securityConnectorsDevOps.json:
```diff
@@ -0,0 +1,3368 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Security/securityConnectors/{securityConnectorName}/devops/default/gitLabGroups/{groupFQName}/listSubgroups": {
```
Is it a typo? Should it be "listSubGroups"?

JohnathonMohr replied:
"Subgroups" is the casing used by GitLab. https://docs.gitlab.com/ee/user/group/subgroups/

ivadim committed:
Update OperationIDs. Update common-types to v5. Update Examples.

ivadim committed:
Update OperationIDs. Update common-types to v5. Update Examples.

ivadim committed:
Resolve linter issues

ivadim commented:
/azp run unifiedPipeline

ivadim committed:
try to reduce operationid

ivadim committed:
Update operations to fix docs generation

ivadim committed:
Update examples

ivadim committed:
Update OperationIDs prefix

ivadim commented:
/azp run

JohnathonMohr committed:
Update to new listAvailable endpoints

JohnathonMohr commented:
/azp run

JohnathonMohr commented:
@ms-henglu Thanks for your review. We have addressed your comments and finished tweaking our API. Please take another look and let us know if you have any further concerns.

JohnathonMohr commented:
@rkmanda looks like you are assigned to this. Can you please take a look at our latest Swagger? Thanks.

mentat9 commented:
>         "x-ms-pageable": {

FYI, this extension doesn't do anything for POST actions. IOW, paging support is not generated in the SDK. It's not harmful to have it here, but it makes sense to remove it.

---
Refers to: specification/security/resource-manager/Microsoft.Security/preview/2023-09-01-preview/securityConnectorsDevOps.json:51 in fa2d657. [](commit_id = fa2d6570d8090220f405a745a5e6c1dba25c6c1a, deletion_comment = False)

mentat9 commented:
>         "x-ms-pageable": {

Has no effect for POST actions. Comment applies to other POST actions in this file.

---
Refers to: specification/security/resource-manager/Microsoft.Security/preview/2023-09-01-preview/securityConnectorsDevOps.json:384 in fa2d657. [](commit_id = fa2d6570d8090220f405a745a5e6c1dba25c6c1a, deletion_comment = False)

mentat9 commented:
>       "description": "Id of the GitHub app installation.",

Would be nice to mention the format (e.g. Uri, Guid, <other>) in the description.

---
Refers to: specification/security/resource-manager/Microsoft.Security/preview/2023-09-01-preview/securityConnectorsDevOps.json:3755 in fa2d657. [](commit_id = fa2d6570d8090220f405a745a5e6c1dba25c6c1a, deletion_comment = False)

mentat9 commented:
@ivadim - Signed off for ARM with comments: please take a look.

JohnathonMohr committed:
Address PR feedback, add ghazdoState property.

JohnathonMohr committed:
Add "ghazdo" (GitHub Advanced Security for Azure DevOps) to custom words.

mentat9 commented:
>         "ghazdoState": {

Recommend only using very common acronyms/abbreviations (e.g. ip, pdf). Maybe githubAdvancedSecurityState...?

---
Refers to: specification/security/resource-manager/Microsoft.Security/preview/2023-09-01-preview/securityConnectorsDevOps.json:3146 in f6d35a2. [](commit_id = f6d35a28777bec8bd0feff91a1e4d60d3ba594e1, deletion_comment = False)

mentat9 commented:
@ivadim - Left a comment, please take a look.

JohnathonMohr committed:
Update 'ghazdo' to 'gitHubAdvancedSecurity'

JohnathonMohr commented:
/azp run unifiedPipeline

ivadim commented:
/azp run

eliagrady reviewed (approved):
LGTM.
Azure-AsyncOperation examples might need to change to reflect concrete API, but this is minor.

eliagrady commented on specification/security/resource-manager/Microsoft.Security/preview/2023-09-01-preview/securityConnectorsDevOps.json:
```diff
@@ -0,0 +1,3785 @@
+        "DeletionFailure"
+      ],
+      "type": "string",
+      "x-ms-enum": {
```
This will conflict with autorest generation with another ProvisioningState already defined:

FATAL: System.InvalidOperationException: Swagger document contains two or more x-ms-enum extensions with the same name 'ProvisioningState' and different values: Succeeded,Failed,Canceled,Pending,PendingDeletion,DeletionSuccess,DeletionFailure vs. Succeeded,Failed,Updating

renaming the definition to ResourceProvisioningState did help to resolve this conflict

JohnathonMohr replied:
Might need input from @ivadim on this.

JohnathonMohr commented on specification/security/resource-manager/Microsoft.Security/preview/2023-09-01-preview/securityConnectorsDevOps.json:
```diff
@@ -0,0 +1,3785 @@
+      "flow": "implicit",
+      "authorizationUrl": "https://login.microsoftonline.com/common/oauth2/authorize",
+      "scopes": {
+        "user_impersonation": "Impersonate your user account"
```
Can you clarify? Will it conflict with other instances in Microsoft.Security?

eliagrady replied:
yes - check the field in other specs to see that it matches exactly in terms of case sensitivity (autorest is picky about that).

JohnathonMohr commented on specification/security/resource-manager/Microsoft.Security/preview/2023-09-01-preview/securityConnectorsDevOps.json:
```diff
@@ -0,0 +1,3785 @@
+            "$ref": "../../../../../common-types/resource-management/v5/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v5/types.json#/parameters/ResourceGroupNameParameter"
```
Will v5 conflict with other endpoints in Microsoft.Security? We updated to v5 because of [this request](https://github.com/Azure/azure-rest-api-specs/pull/24384#discussion_r1234777022) on this review. If using v5 will cause AutoRest to fail because of usage of v2 elsewhere in Microsoft.Security, we need to communicate that as the reason for not following the guidance.

eliagrady replied:
v2 in my PR is v5 copied with alternate descriptions - a move to v5 will have to be coordinated in a new, cross MDC API version

JohnathonMohr committed:
Updating with latest API design, bringing more consistency to descriptions, addressing feedback from Microsoft.Security RP.

JohnathonMohr committed:
Fix Azure DevOps Repositories examples

eliagrady commented on specification/security/resource-manager/Microsoft.Security/preview/2023-09-01-preview/securityConnectorsDevOps.json:
```diff
@@ -0,0 +1,3785 @@
+    "SecurityConnectorName": {
+      "in": "path",
+      "name": "securityConnectorName",
+      "description": "Name of the Security Connector.",
```
the description should change

eliagrady reviewed (approved):
approved with suggestions - this version will have to change a bit for a successful SDK gen with autorest

ivadim committed:
Rename Configurations and OperationResults to all SDK build

ivadim committed:
Update examples

ivadim committed:
Allign examples with MDC

ivadim commented:
/pr RequestMerge

ivadim commented:
On Linter Issues:
1.  PutResponseSchemaDescription -> looks like all false positive, We have Create or Update in description.
2. LroLocationHeader -> false positive, we use Azure-AsyncOperation header by ResourceProvider specification
3.  GetCollectionOnlyHasValueAndNextLink -> false positive, it points to the GET request to "singletone" object, there is a separate list for the parent
4. ResourceNameRestriction -> we can't fully control names of the resources because they comes from external sources (GitHub, GitLab, AzureDevOps) and they do not always post restriction patterns.

rkmanda commented:
> On Linter Issues:
> 
> 1. PutResponseSchemaDescription -> looks like all false positive, We have Create or Update in description.
> 2. LroLocationHeader -> false positive, we use Azure-AsyncOperation header by ResourceProvider specification
> 3. GetCollectionOnlyHasValueAndNextLink -> false positive, it points to the GET request to "singletone" object, there is a separate list for the parent
> 4. ResourceNameRestriction -> we can't fully control names of the resources because they comes from external sources (GitHub, GitLab, AzureDevOps) and external vendors do not always post restriction patterns.

You may apply suppressions for all these. How about "[RepeatedPathInfo](https://github.com/Azure/azure-openapi-validator/blob/main/docs/repeated-path-info.md)"

https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process

rkmanda commented:
Please add the MergeRequested label again once you have addressed the comments

ivadim committed:
Fix and suppress linter issues

ivadim commented:
@rkmanda 
> Please add the MergeRequested label again once you have addressed the comments

If fixed RepeatedPathInfo and PutResponseSchemaDescription . Added others to suppressions list. 

Could you please approve the suppressions?

ivadim committed:
use new suppression syntax

ivadim committed:
duplicate supressions

ivadim commented:
/pr RequestMerge

ivadim committed:
fix typo in description
